src/environments/vumps_environments.jl

```diff
@@ -0,0 +1,516 @@
+a scalar factor `λ` such that ``λ*AL*L = L*A``, where an initial guess for `L` can be
+provided.
+"""
+function left_canonical(A::Matrix{<:AbstractTensorMap}, L::Matrix{<:AbstractTensorMap} = initial_C(A); tol = 1e-12, maxiter = 100, kwargs...)
+    L = getL!(A, L; kwargs...)
+    AL, Le, λ = getAL(A, L;kwargs...)
+    numiter = 1
+    while norm(L.-Le) > tol && numiter < maxiter
+        L = getLsped(Le, A, AL; kwargs...)
+        AL, Le, λ = getAL(A, L; kwargs...)
+        numiter += 1
+    end
+    L = Le
+    return AL, L, λ
+end
+
+"""
+    R, AR, λ = right_canonical(A::Matrix{<:AbstractTensorMap}, L::Matrix{<:AbstractTensorMap} = initial_C(A); tol = 1e-12, maxiter = 100, kwargs...)
+
+Given an MPS tensor `A`, return a gauge transform R, a right-canonical MPS tensor `AR`, and
+a scalar factor `λ` such that ``λ * R * AR = A * R``, where an initial guess for `R` can be
+provided.
+"""
+function right_canonical(A::Matrix{<:AbstractTensorMap}, L::Matrix{<:AbstractTensorMap} = initial_C(A); tol = 1e-12, maxiter = 100, kwargs...)
+    Ar = [permute(A, ((4,2,3), (1,))) for A in A]
+    Lr = [permute(L, ((2,   ), (1,))) for L in L]
+    
+    AL, L, λ = left_canonical(Ar, Lr; tol, maxiter, kwargs...)
+
+     R = [permute( L, ((2,), (1,   ))) for  L in  L]
+    AR = [permute(AL, ((4,2,3), (1,))) for AL in AL]
+    return R, AR, λ
+end
+
+function initial_FL(AL::Matrix{<:AbstractTensorMap}, ipeps::InfinitePEPS)
+    T = eltype(ipeps[1])
+    FL = [(D = space(ipeps, 5)';
+          χ = space(AL, 4)';
+          TensorMap(rand, T, χ * D * D', χ)) for (ipeps, AL) in zip(ipeps.A, AL)]
+    return FL
+end
+
+function initial_FR(AR::Matrix{<:AbstractTensorMap}, ipeps::InfinitePEPS)
+    T = eltype(ipeps[1])
+    FR = [(D = space(ipeps, 3)';
+           χ = space(AR, 4)';
+           TensorMap(rand, T, χ * D * D', χ)) for (ipeps, AR) in zip(ipeps.A, AR)]
+    return FR
+end
+
+"""
+    λL, FL = leftenv(ALu, ALd, O, FL = initial_FL(ALu,O); kwargs...)
+
+Compute the left environment tensor for MPS A and MPO O, by finding the left fixed point
+of ALu - O - ALd contracted along the physical dimension.
+```
+ ┌──  ALuᵢⱼ  ──          ┌── 
+ │     │                 │   
+FLᵢⱼ ─ Oᵢⱼ   ──   = λLᵢⱼ FLᵢⱼ₊₁   
+ │     │                 │   
+ └──  ALdᵢᵣⱼ  ─          └── 
+```
+"""
+function leftenv(ALu::Matrix{<:AbstractTensorMap}, 
+        ALd::Matrix{<:AbstractTensorMap}, 
+        ipeps::InfinitePEPS, 
+        FL::Matrix{<:AbstractTensorMap} = initial_FL(ALu,ipeps); 
+        ifobs=false, verbosity = Defaults.verbosity, kwargs...) 
+
+    Ni, Nj = size(ipeps)
+    λL = Zygote.Buffer(zeros(eltype(ipeps[1]), Ni))
+    FL′ = Zygote.Buffer(FL)
+    for i in 1:Ni
+        ir = ifobs ? Ni + 1 - i : mod1(i + 1, Ni)
+        λLs, FL1s, info = eigsolve(FLi -> FLmap(FLi, ALu[i,:], ALd[ir,:], ipeps.A[i,:], adjoint.(ipeps.A[i,:])), 
+                                   FL[i,:], 1, :LM; maxiter=100, ishermitian = false, kwargs...)
```

Comment:
I previously examined the speed and convergence. From my observations, they appear to be identical. However, I prefer this implementation because it processes the entire environment simultaneously.